nvflare/apis/utils/reliable_message.py

```diff
@@ -0,0 +1,369 @@
+                return make_reply(ReturnCode.TASK_ABORTED)
+
+            ack = engine.send_aux_request(
+                targets=[target],
+                topic=TOPIC_RELIABLE_REQUEST,
+                request=request,
+                timeout=timeout,
+                fl_ctx=fl_ctx,
+            )
+            ack, rc = _extract_result(ack, target)
+            if ack and rc != ReturnCode.COMMUNICATION_ERROR:
+                # is this result?
+                op = ack.get_header(HEADER_OP)
+                if op == OP_REPLY:
+                    # the reply is already the result - we are done!
+                    # this could happen when we didn't get positive ack for our first request, and the result was
+                    # already produced when we did the 2nd request (this request).
+                    return ack
+
+                # the ack is a status report - check status
+                status = ack.get_header(HEADER_STATUS)
+                if status and status != STATUS_NOT_RECEIVED:
+                    # status should never be STATUS_NOT_RECEIVED, unless there is a bug in the receiving logic
+                    # STATUS_NOT_RECEIVED is only possible during "query" phase.
+                    break
+
+            # we didn't get a positive ack - wait a short time and re-send the request.
+            num_tries += 1
+            if num_tries > cls._max_retries:
+                # enough tries
+                return _error_reply(ReturnCode.COMMUNICATION_ERROR, f"Max send retries ({cls._max_retries}) reached")
+            time.sleep(cls._query_interval)
+
+        # Querying phase - try to get result
+        query = Shareable()
+        query.set_header(HEADER_TX, receiver.tx_id)
+        query.set_header(HEADER_OP, OP_QUERY)
+
+        num_tries = 0
+        while True:
+            if receiver.result_ready.wait(cls._query_interval):
+                # we already received result sent by the target.
+                # Note that we don't wait forever here - we only wait for _query_interval so we could
+                # check other condition and/or send query to ask for result.
+                return receiver.result
+
+            if abort_signal.triggered:
+                return make_reply(ReturnCode.TASK_ABORTED)
+
+            # send a query. The ack of the query could be the result itself, or a status report.
+            # Note: the ack could be the result because we failed to receive the result sent by the target earlier.
+            ack = engine.send_aux_request(
+                targets=[target],
+                topic=TOPIC_RELIABLE_REQUEST,
+                request=query,
+                timeout=timeout,
+                fl_ctx=fl_ctx,
+            )
+            ack, rc = _extract_result(ack, target)
+            if ack and rc != ReturnCode.COMMUNICATION_ERROR:
+                op = ack.get_header(HEADER_OP)
+                if op == OP_REPLY:
+                    # the ack is result itself!
```

Comment:
Do you think it make sense to have separate methods 
- one call "_send_request" which contains the previous while loop
- and one method call "_query_result" which contains this second while loop

Reply:
Not sure what you  mean, but this is to try to make it more efficient by reducing the number of messages. So the query reply could be the result, or could also be an ack.

Reply:
I am suggesting we break this method into 2 smaller methods, each has their own for loop.

But that is just personal preference for better readability, logic wise this part LGTM

Reply:
Good idea. Changed.